probe-rs/src/probe/blackmagic/arm.rs

```diff
@@ -0,0 +1,815 @@
+                rnw: 0,
+                addr,
+                value,
+            },
+            ProtocolVersion::V1 | ProtocolVersion::V2 => RemoteCommand::RawAccessV1 {
+                index,
+                rnw: 0,
+                addr,
+                value,
+            },
+            ProtocolVersion::V3 | ProtocolVersion::V4 => RemoteCommand::RawAccessV3 {
+                index,
+                rnw: 0,
+                addr,
+                value,
+            },
+        };
+        let result = self
+            .probe
+            .command(command)
+            .map_err(|e| ArmError::Probe(e.into()))?
+            .0;
+        if result == 0 {
+            Ok(())
+        } else {
+            Err(ArmError::Probe(DebugProbeError::Other(format!(
+                "probe returned unexpected result: {}",
+                result
+            ))))
+        }
+    }
+
+    fn read_raw_ap_register(
+        &mut self,
+        ap: &FullyQualifiedApAddress,
+        addr: u8,
+    ) -> Result<u32, ArmError> {
+        let (index, apsel) = ap_to_bmp(ap)?;
+
+        let command = match self.probe.remote_protocol {
+            ProtocolVersion::V0 => {
+                return Err(ArmError::Probe(
+                    DebugProbeError::CommandNotSupportedByProbe {
+                        command_name: "adiv5 raw ap read",
+                    },
+                ));
+            }
+            ProtocolVersion::V0P => RemoteCommand::ReadApV0P { apsel, addr },
+            ProtocolVersion::V1 | ProtocolVersion::V2 => {
+                RemoteCommand::ReadApV1 { index, apsel, addr }
+            }
+            ProtocolVersion::V3 | ProtocolVersion::V4 => {
+                RemoteCommand::ReadApV3 { index, apsel, addr }
+            }
+        };
+        let result = u32::from_be(
+            TryInto::<u32>::try_into(
+                self.probe
+                    .command(command)
+                    .map_err(|e| ArmError::Probe(e.into()))?
+                    .0,
+            )
+            .unwrap(),
+        );
+        Ok(result)
+    }
+
+    fn write_raw_ap_register(
+        &mut self,
+        ap: &FullyQualifiedApAddress,
+        addr: u8,
+        value: u32,
+    ) -> Result<(), ArmError> {
+        let (index, apsel) = ap_to_bmp(ap)?;
+        let command = match self.probe.remote_protocol {
+            ProtocolVersion::V0 => {
+                return Err(ArmError::Probe(
+                    DebugProbeError::CommandNotSupportedByProbe {
+                        command_name: "adiv5 raw ap write",
+                    },
+                ));
+            }
+            ProtocolVersion::V0P => RemoteCommand::WriteApV0P { apsel, addr, value },
+            ProtocolVersion::V1 | ProtocolVersion::V2 => RemoteCommand::WriteApV1 {
+                index,
+                apsel,
+                addr,
+                value,
+            },
+            ProtocolVersion::V3 | ProtocolVersion::V4 => RemoteCommand::WriteApV3 {
```

Comment:
Is there no `WriteApV2`? (just curious how the mappings came to happen :))

Reply:
The mappings are based on hapazard protocol extensions to the Black Magic Probe protocol. V2 did not change from V1, so it just uses the previous version.

This was largely compiled by referencing existing implementations at https://github.com/blackmagic-debug/blackmagic/tree/main/src/platforms/hosted/remote

Reply:
Oki, nice :) Thanks!